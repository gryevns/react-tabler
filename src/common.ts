export type Color =
    | "blue"
    | "azure"
    | "indigo"
    | "purple"
    | "pink"
    | "red"
    | "orange"
    | "yellow"
    | "lime"
    | "green"
    | "teal"
    | "cyan"
    | "gray";

export const colors = [
    "blue" as const,
    "azure" as const,
    "indigo" as const,
    "purple" as const,
    "pink" as const,
    "red" as const,
    "orange" as const,
    "yellow" as const,
    "lime" as const,
    "green" as const,
    "teal" as const,
    "cyan" as const,
    "gray" as const,
];

export type Variant =
    | "primary"
    | "secondary"
    | "success"
    | "info"
    | "warning"
    | "danger";

export const variants = [
    "primary" as const,
    "secondary" as const,
    "success" as const,
    "info" as const,
    "warning" as const,
    "danger" as const,
];

export type Size = "sm" | "md" | "lg" | "xl";

export const sizes = ["sm" as const, "md" as const, "lg" as const, "xl" as const];
