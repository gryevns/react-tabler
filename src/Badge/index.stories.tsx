import React from "react";

import { colors } from "../common";
import Badge from "../Badge";

export default {
    title: "Badge",
};

export const Colors = () =>
    colors.map(color => (
        <Badge color={color} className="mr-1">
            {color}
        </Badge>
    ));

export const Pill = () =>
    colors.map(color => (
        <Badge color={color} pill className="mr-1">
            {color}
        </Badge>
    ));

export const Soft = () =>
    colors.map(color => (
        <Badge color={color} soft className="mr-1">
            {color}
        </Badge>
    ));

export const Link = () =>
    colors.map(color => (
        <Badge link="#" color={color} className="mr-1">
            {color}
        </Badge>
    ));

export const Empty = () =>
    colors.map(color => <Badge color={color} className="mr-1" />);

export const Addons = () => (
    <>
        <Badge color="green">
            Task<Badge.Addon>Finished</Badge.Addon>
        </Badge>
        <Badge color="purple">
            bundle<Badge.Addon>passing</Badge.Addon>
        </Badge>
        <Badge color="red" soft>
            XSS gzip size<Badge.Addon color="red">20.9kB</Badge.Addon>
        </Badge>
    </>
);

export const Avatar = () => (
    <>
        <Badge color="blue">
            <Badge.Avatar imageURL="https://picsum.photos/id/353/128/128">
                JL
            </Badge.Avatar>
            Pawe≈Ç Kuna
        </Badge>
        <Badge color="blue">
            <Badge.Avatar>JL</Badge.Avatar>
            Jeffie Lewzey
        </Badge>
    </>
);
