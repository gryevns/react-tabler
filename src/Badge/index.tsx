import * as React from "react";
import cn from "classnames";
import { Color, Variant } from "../common";

export interface AddonProps {
    children: React.ReactNode;
    color?: Color;
}

const Addon = ({ children, color }: AddonProps) => {
    const classes = cn("badge-addon", { [`bg-${color}`]: color });
    return <span className={classes}>{children}</span>;
};

export interface AvatarProps {
    children?: React.ReactNode;
    imageURL?: string;
    style?: React.CSSProperties;
}

const Avatar = ({ children, imageURL, style = {} }: AvatarProps) => (
    <span
        className="badge-avatar"
        style={{ ...style, backgroundImage: `url(${imageURL})` }}
    >
        {children}
    </span>
);

export interface BadgeProps {
    children?: React.ReactNode;
    className?: string;
    pill?: boolean;
    soft?: boolean;
    style?: React.CSSProperties;
    color?: Color;
    link?: string;
    variant?: Variant;
}

export const Badge = ({
    className,
    children,
    pill = false,
    soft = false,
    style,
    color,
    variant,
    link = null,
}: BadgeProps) => {
    const classes = cn(
        {
            badge: true,
            "badge-pill": pill,
            [`bg-${variant}`]: !!variant,
            [!soft ? `bg-${color}` : `bg-${color}-lt`]: color,
        },
        className,
    );
    if (link) {
        return (
            <a className={classes} href={link}>
                {children}
            </a>
        );
    }
    return (
        <span style={style} className={classes}>
            {children}
        </span>
    );
};

Badge.Addon = Addon;
Badge.Avatar = Avatar;

export default Badge;
