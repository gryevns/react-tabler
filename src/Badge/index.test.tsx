import React from "react";
import { shallow } from "enzyme";

import { colors, Color } from "../common";
import Badge from "./index";

describe("Badge", () => {
    it("should match color snapshot", () => {
        colors.map((color: Color) => {
            const wrapper = shallow(<Badge color={color} />);
            expect(wrapper).toMatchSnapshot();
        });
    });
    it("should match pill snapshot", () => {
        const wrapper = shallow(<Badge pill />);
        expect(wrapper).toMatchSnapshot();
    });
    it("should match soft snapshot", () => {
        const wrapper = shallow(<Badge soft />);
        expect(wrapper).toMatchSnapshot();
    });
    it("should match link snapshot", () => {
        const wrapper = shallow(<Badge link="https://www.test.com" />);
        expect(wrapper).toMatchSnapshot();
    });
    it("should match empty snapshot", () => {
        const wrapper = shallow(<Badge />);
        expect(wrapper).toMatchSnapshot();
    });
    it("should match addon snapshot", () => {
        const wrapper = shallow(
            <Badge color="green">
                Task<Badge.Addon>Finished</Badge.Addon>
            </Badge>,
        );
        expect(wrapper).toMatchSnapshot();
    });
    it("should match avatar snapshot", () => {
        const wrapper = shallow(
            <Badge color="blue">
                <Badge.Avatar imageURL="https://picsum.photos/id/353/128/128">
                    JL
                </Badge.Avatar>
                ABC DEF
            </Badge>,
        );
        expect(wrapper).toMatchSnapshot();
    });
});
