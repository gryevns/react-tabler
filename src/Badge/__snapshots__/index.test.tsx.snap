// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Badge should match addon snapshot 1`] = `
<span
  className="badge bg-green"
>
  Task
  <Addon>
    Finished
  </Addon>
</span>
`;

exports[`Badge should match avatar snapshot 1`] = `
<span
  className="badge bg-blue"
>
  <Avatar
    imageURL="https://picsum.photos/id/353/128/128"
  >
    JL
  </Avatar>
  ABC DEF
</span>
`;

exports[`Badge should match color snapshot 1`] = `
<span
  className="badge bg-blue"
/>
`;

exports[`Badge should match color snapshot 2`] = `
<span
  className="badge bg-azure"
/>
`;

exports[`Badge should match color snapshot 3`] = `
<span
  className="badge bg-indigo"
/>
`;

exports[`Badge should match color snapshot 4`] = `
<span
  className="badge bg-purple"
/>
`;

exports[`Badge should match color snapshot 5`] = `
<span
  className="badge bg-pink"
/>
`;

exports[`Badge should match color snapshot 6`] = `
<span
  className="badge bg-red"
/>
`;

exports[`Badge should match color snapshot 7`] = `
<span
  className="badge bg-orange"
/>
`;

exports[`Badge should match color snapshot 8`] = `
<span
  className="badge bg-yellow"
/>
`;

exports[`Badge should match color snapshot 9`] = `
<span
  className="badge bg-lime"
/>
`;

exports[`Badge should match color snapshot 10`] = `
<span
  className="badge bg-green"
/>
`;

exports[`Badge should match color snapshot 11`] = `
<span
  className="badge bg-teal"
/>
`;

exports[`Badge should match color snapshot 12`] = `
<span
  className="badge bg-cyan"
/>
`;

exports[`Badge should match color snapshot 13`] = `
<span
  className="badge bg-gray"
/>
`;

exports[`Badge should match empty snapshot 1`] = `
<span
  className="badge"
/>
`;

exports[`Badge should match link snapshot 1`] = `
<a
  className="badge"
  href="https://www.test.com"
/>
`;

exports[`Badge should match pill snapshot 1`] = `
<span
  className="badge badge-pill"
/>
`;

exports[`Badge should match soft snapshot 1`] = `
<span
  className="badge"
/>
`;
